### PR TITLE
KOGITO-8062 DSL: Refactor environments and export to config

### PR DESCRIPTION
**Pull request by radtriste**
No Environment class anymore.
Nightly and Pull Request folders will be created on the fly for the specified environments with the configuration given into the yaml file.
That way, we can have multiple project, using the same folder structure but with different environments.

Related PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/622

- https://github.com/kiegroup/drools/pull/4710
- https://github.com/kiegroup/drools-website/pull/231

- https://github.com/kiegroup/kogito-runtimes/pull/2535
- https://github.com/kiegroup/kogito-apps/pull/1490
- https://github.com/kiegroup/kogito-examples/pull/1441
- https://github.com/kiegroup/kogito-images/pull/1350
- https://github.com/kiegroup/kogito-operator/pull/1306
- https://github.com/kiegroup/kogito-serverless-operator/pull/44
- https://github.com/kiegroup/kogito-docs/pull/230

- https://github.com/kiegroup/kie-tools/pull/1279

### CODE REVIEW

File: dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch [hunk 1]
```diff
@@ -107,6 +107,9 @@ pipeline {
                     all_repos.each { repository ->
                         def envProps = getRepoConfigAsEnvProperties(repository)
                         envProps += [
+                            DEBUG: scriptUtils.isDebug(),
+                            ENVIRONMENTS: getRepoConfig(repository).environments ? getRepoConfig(repository).environments.keySet().join(',') : '',
```
Comment (Ginxo) [inline dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch]:
```suggestion
                            ENVIRONMENTS: getRepoConfig(repository).environments?.keySet().join(',') ?: '',
```
or even
```
                            ENVIRONMENTS: getRepoConfig(repository)?.environments?.keySet()?.join(',') ?: ''
```

See 
image

File: dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch [hunk 2]
```diff
@@ -295,8 +298,8 @@ void fillEnvProperties(Map envProperties, String envKeyPrefix, Map propsMap) {
         }
         if (value instanceof Map) {
             fillEnvProperties(envProperties, newKey, value as Map)
-        } else if (value instanceof List) {
-            envProperties.put(newKey, (value as List).join(','))
+        } else if (value instanceof List || value instanceof net.sf.json.JSONArray) {
```
Comment (Ginxo) [inline dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch]: a proposal for making it more maintainable (but yours is fine too, so up to you)
 ```suggestion
        } else if ([List, net.sf.json.JSONArray].any { it.isAssignableFrom(value.getClass()) }) {
```
See
image

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy [hunk 3]
```diff
@@ -39,22 +44,22 @@ class KogitoJobTemplate {
         String jobFolderName = ''
         Map jobFolderEnv = [:]
         if (jobParams.job.folder) {
-            if (!jobParams.job.folder instanceof Folder) {
-                throw new RuntimeException('Folder is not of type org.kie.jenkins.jobdsl.model.Folder')
+            if (!jobParams.job.folder instanceof Folder && !jobParams.job.folder instanceof JenkinsFolder) {
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]: same proposal here
```suggestion
            if (![Folder, JenkinsFolder].any { it.isAssignableFrom(jobParams.job.folder.getClass()) }) {
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy [hunk 4]
```diff
@@ -292,9 +317,22 @@ class KogitoJobTemplate {
...
+        String testTypeId
+        String testTypeName
+        if (prFolder instanceof Folder) {
+            // DEPRECATED
+            // Should be deleted once https://issues.redhat.com/browse/PLANNER-2870 is implemented
+            testTypeId = prFolder.getConfigValues()?.typeId ?: prFolder.environment.toId()
+            testTypeName = prFolder.getConfigValues()?.typeName ?: prFolder.environment.toId()
+        } else if (prFolder instanceof JenkinsFolder) {
+            String envName
+            testTypeId = envName ? envName : 'tests' // Define harcoded value for no env
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]:
```suggestion
            testTypeId = envName ?: 'tests' // Define harcoded value for no env
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy [hunk 5]
```diff
@@ -292,9 +317,22 @@ class KogitoJobTemplate {
...
+            testTypeName = envName ? envName : 'build' // Define harcoded value for no env
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]:
```suggestion
            testTypeName = envName ?: 'build' // Define harcoded value for no env
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/Folder.groovy [hunk 6]
```diff
@@ -14,7 +20,11 @@ class Folder {
...
     String getFolderName(String separator = '.') {
-        String folderName = this.jobType.toName()
+        return getName(separator)
+    }
+
+    String getName(String separator = '.') {
+        String folderName = this.jobType.getName()
         if (this.environment != Environment.DEFAULT) {
             folderName += "${separator}${environment.toId()}"
         }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/Folder.groovy:30]:
```suggestion
    String getName(String separator = '.') {
        return "${this.jobType.getName()}${this.environment != Environment.DEFAULT ? "${separator}${environment.toId()}" : ''}"
```
:thinking: and to remove next `return` line (I can't propose it from the UI)

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolder.groovy [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    */
+    Map getDefaultEnvVars(def script) {
+        return getDefaultEnvVars()
+    }
+    Map getDefaultEnvVars() {
+        Map env = [:]
+        env.putAll(this.jobType.getDefaultEnvVars() ?: [:])
+        env.putAll(this.defaultEnv ?: [:])
+        return new HashMap(env)
+    }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolder.groovy:51]:
```suggestion
    Map getDefaultEnvVars() {
        return new HashMap(this.jobType.getDefaultEnvVars() ?: [:] + this.defaultEnv ?: [:])
    }
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            defaultEnv = EnvUtils.getEnvironmentEnvVars(script, envName)
+        }
+        JenkinsFolder folder = new JenkinsFolder(jobType, defaultEnv, envName)
+
+        if (hasFolder(folder.getName())) {
+            return getFolder(folder.getName())
+        }
+
+        register(folder)
+        return folder
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy]:
```suggestion
        if (hasFolder(folder.getName())) {
            return getFolder(folder.getName())
        } else {
            register(folder)
            return folder
        }
```
to make readability easier and to have one single exit point per block

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    static JenkinsFolder getFolder(String folderName) {
+        if (hasFolder(folderName)) {
+            return FOLDER_REGISTRY.get(folderName)
+        }
+        return null
+    }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy]:
```suggestion
    static JenkinsFolder getFolder(String folderName) {
        return hasFolder(folderName) ? FOLDER_REGISTRY.get(folderName) : null
    }
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    static List<String> getAllEnabledEnvironments(def script) {
+        List<String> environments = getAllEnvironments(script)
+        List<String> enabledEnvironments = []
+        for (String envName : environments) {
+            if (isEnvironmentEnabled(script, envName)) {
+                enabledEnvironments += envName
+            }
+        }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]:
```suggestion
        List<String> enabledEnvironments = environments.findAll{ isEnvironmentEnabled(script, it) }
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    static List<String> getAllEnabledEnvironmentsForIds(def script, List<String> includeEnvironmentIds, List<String> excludeEnvironmentIds) {
+        List<String> environments = getAllEnabledEnvironments(script)
+        List<String> filteredEnvironments = []
+        for (String envName : environments) {
+            if (hasEnvironmentIds(script, envName, includeEnvironmentIds) &&
+                !hasEnvironmentIds(script, envName, excludeEnvironmentIds)) {
+                filteredEnvironments += envName
+            }
+        }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]:
```suggestion
        List<String> filteredEnvironments = environments.findAll{ hasEnvironmentIds(script, it, includeEnvironmentIds) && !hasEnvironmentIds(script, it, excludeEnvironmentIds) }        
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        String keyPrefix = "${createEnvironmentsKeyPrefix(envName)}ENV_VARS_"
+        PrintUtils.debug(script, "Looking for env vars for env ${envName}. Key prefix is ${keyPrefix}")
+        Map envVars = [:]
+        for (String key : script.getBinding().getVariables().keySet()) {
+            if (key.startsWith(keyPrefix)) {
+                String envKey = key.replace(keyPrefix, '')
+                String envValue = Utils.getBindingValue(script, key)
+                envVars.put(envKey, envValue)
+            }
+        }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]:
```suggestion
        Map envVars =script.getBinding().getVariables().keySet()
                                      .findAll{ it.startsWith(keyPrefix) } 
                                      .collectEntries{ [it.replace(keyPrefix, ''), Utils.getBindingValue(script, it)] } 
```

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    static boolean hasEnvironmentIds(def script, String envName, List<String> ids) {
+        for (String id : ids) {
+            if (!hasEnvironmentId(script, envName, id)) {
+                return false
+            }
+        }
+        return true
+    }
```
Comment (Ginxo) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]:
```suggestion
    static boolean hasEnvironmentIds(def script, String envName, List<String> ids) {
        return !ids.any{ !hasEnvironmentId(script, envName, it) }
    }
```

**Review by Ginxo — APPROVED** [review]
I'm ok with the change, just few technical proposals

File: dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch (hunk 1 above)
Comment (radtriste) [reply dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch]: gonna use `getRepoConfig(repository).environments?.keySet()?.join(',') ?: '',`

File: dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch (hunk 2 above)
Comment (radtriste) [reply dsl/seed/jenkinsfiles/Jenkinsfile.seed.branch]: taking it !

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy (hunk 3 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]: taking it

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy (hunk 4 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]: good catch

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy (hunk 5 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/KogitoJobTemplate.groovy]: good catch

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/Folder.groovy (hunk 6 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/Folder.groovy:30]: This is meant to disappear once I update optaplanner with new JenkinsFolder.
And I find it more readable this way :)

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolder.groovy (hunk 7 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolder.groovy:51]: same here, having more lines makes it more readable IMO

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy (hunk 8 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy]: updated

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy (hunk 9 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/model/JenkinsFolderRegistry.groovy]: updated

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy (hunk 10 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]: updated

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy (hunk 11 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]: updated

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy (hunk 12 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]: thx ! updated

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy (hunk 13 above)
Comment (radtriste) [reply dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/utils/EnvUtils.groovy]: using `.every` instead

**Review by Ginxo — APPROVED** [review]
LGTM

**Pull request merged**